// @storybook/react/dist/server/framework-preset-cra.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webpackFinal = webpackFinal;
exports.babelDefault = babelDefault;

var _nodeLogger = require("@storybook/node-logger");

var _craConfig = require("./cra-config");

function webpackFinal(config, {
  configDir
}) {
  if (!(0, _craConfig.isReactScriptsInstalled)()) {
    _nodeLogger.logger.info('=> Using base config because react-scripts is not installed.');

    return config;
  }

  _nodeLogger.logger.info('=> Loading create-react-app config.');

  return (0, _craConfig.applyCRAWebpackConfig)(config, configDir);
}

function babelDefault(config) {
  if (!(0, _craConfig.isReactScriptsInstalled)()) {
    return config;
  }

  return Object.assign({}, config, {
    presets: [require.resolve('babel-preset-react-app')],
    plugins: [[require.resolve('babel-plugin-named-asset-import'), {
      loaderMap: {
        svg: {
          ReactComponent: '@svgr/webpack?-prettier,-svgo![path]'
        }
      }
    }]]
  });
}
