// @storybook/ui/dist/components/preview/components.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnstyledLink = exports.FrameWrap = exports.Frame = void 0;

var _theming = require("@storybook/theming");

var _router = require("@storybook/router");

var Frame = _theming.styled.div({
  position: 'absolute',
  top: 0,
  left: 0,
  border: '0 none',
  transition: 'transform .2s ease-out, height .2s ease-out, width .2s ease-out',
  transformOrigin: 'top left',
  overflow: 'auto',
  '& > iframe': {
    width: '100%',
    height: '100%',
    position: 'absolute',
    top: 0,
    left: 0
  }
});

exports.Frame = Frame;

var FrameWrap = _theming.styled.div(function (_ref) {
  var offset = _ref.offset;
  return {
    position: 'absolute',
    overflow: 'auto',
    left: 0,
    right: 0,
    bottom: 0,
    top: offset,
    zIndex: 3,
    transition: 'all 0.1s linear',
    height: "calc(100% - ".concat(offset, "px)"),
    background: 'transparent'
  };
});

exports.FrameWrap = FrameWrap;
var UnstyledLink = (0, _theming.styled)(_router.Link)({
  color: 'inherit',
  textDecoration: 'inherit',
  display: 'inline-block'
});
exports.UnstyledLink = UnstyledLink;
