// @storybook/components/dist/tooltip/TooltipLinkList.stories.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.links = void 0;

var _react = _interopRequireDefault(require("react"));

var _react2 = require("@storybook/react");

var _addonActions = require("@storybook/addon-actions");

var _WithTooltip = _interopRequireDefault(require("./WithTooltip"));

var _TooltipLinkList = _interopRequireDefault(require("./TooltipLinkList"));

var _StoryLinkWrapper = _interopRequireDefault(require("../StoryLinkWrapper"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var links = [{
  id: '1',
  title: 'Link',
  href: 'http://google.com'
}, {
  id: '2',
  title: 'Link',
  href: 'http://google.com'
}, {
  id: '3',
  title: 'callback',
  onClick: (0, _addonActions.action)('onClick')
}];
exports.links = links;

var _ref =
/*#__PURE__*/
_react.default.createElement("div", null, "Tooltip");

var _ref2 =
/*#__PURE__*/
_react.default.createElement(_TooltipLinkList.default, {
  links: links,
  LinkWrapper: _StoryLinkWrapper.default
});

(0, _react2.storiesOf)('basics/Tooltip/TooltipLinkList', module).addDecorator(function (storyFn) {
  return _react.default.createElement("div", {
    style: {
      height: '300px'
    }
  }, _react.default.createElement(_WithTooltip.default, {
    placement: "top",
    trigger: "click",
    tooltipShown: true,
    tooltip: storyFn()
  }, _ref));
}).add('links', function () {
  return _react.default.createElement(_TooltipLinkList.default, {
    links: links.slice(0, 2),
    LinkWrapper: _StoryLinkWrapper.default
  });
}).add('links and callback', function () {
  return _ref2;
});
