// @storybook/ui/dist/core/channel.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _utilDeprecate = _interopRequireDefault(require("util-deprecate"));

var _coreEvents = require("@storybook/core-events");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = function _default(_ref) {
  var provider = _ref.provider;
  var api = {
    getChannel: function getChannel() {
      return provider.channel;
    },
    on: function on(type, cb) {
      var peer = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

      if (peer) {
        provider.channel.addPeerListener(type, cb);
      } else {
        provider.channel.addListener(type, cb);
      }

      return function () {
        return provider.channel.removeListener(type, cb);
      };
    },
    off: function off(type, cb) {
      provider.channel.removeListener(type, cb);
    },
    emit: function emit(type, event) {
      provider.channel.emit(type, event);
    },
    onStory: (0, _utilDeprecate.default)(function (cb) {
      return api.on(_coreEvents.STORY_CHANGED, cb);
    }, 'onStory(...) has been replaced with on(STORY_CHANGED, ...)')
  };
  return {
    api: api
  };
};

exports.default = _default;
