// @storybook/addon-actions/dist/preview/action.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = action;

require("core-js/modules/es6.object.assign");

var _v = _interopRequireDefault(require("uuid/v1"));

var _addons = _interopRequireDefault(require("@storybook/addons"));

var _constants = require("../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function action(name) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var actionOptions = Object.assign({}, options); // eslint-disable-next-line no-shadow

  var handler = function action() {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    var channel = _addons.default.getChannel();

    var id = (0, _v.default)();
    channel.emit(_constants.EVENT_ID, {
      id: id,
      data: {
        name: name,
        args: args
      },
      options: actionOptions
    });
  };

  return handler;
}
