// @storybook/addon-actions/dist/components/ActionLogger/style.js
"use strict";

require("core-js/modules/es6.object.define-property");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Wrapper = exports.InspectorContainer = exports.Counter = exports.Action = exports.Actions = void 0;

require("core-js/modules/es6.string.bold");

var _theming = require("@storybook/theming");

var _polished = require("polished");

var Actions = _theming.styled.pre({
  flex: 1,
  margin: 0,
  padding: '10px 5px 20px',
  overflowY: 'auto',
  color: '#666'
});

exports.Actions = Actions;

var Action = _theming.styled.div({
  display: 'flex',
  padding: '0',
  borderLeft: '5px solid transparent',
  borderBottom: '1px solid transparent',
  transition: 'all 0.1s',
  alignItems: 'flex-start'
});

exports.Action = Action;

var Counter = _theming.styled.div(function (_ref) {
  var theme = _ref.theme;
  return {
    backgroundColor: (0, _polished.opacify)(0.5, theme.appBorderColor),
    color: theme.color.inverseText,
    fontSize: theme.typography.size.s1,
    fontWeight: theme.typography.weight.bold,
    lineHeight: 1,
    padding: '1px 5px',
    borderRadius: '20px',
    margin: '2px 0px'
  };
});

exports.Counter = Counter;

var InspectorContainer = _theming.styled.div({
  flex: 1,
  padding: '0 0 0 5px'
});

exports.InspectorContainer = InspectorContainer;

var Wrapper = _theming.styled.div({
  flex: 1,
  display: 'flex',
  position: 'relative',
  height: '100%'
});

exports.Wrapper = Wrapper;
